lib/cli/src/commands/run/capabilities/net.rs

```diff
@@ -0,0 +1,306 @@
+use std::{
+    fmt::Display,
+    net::{IpAddr, SocketAddr},
+    path::PathBuf,
+    str::FromStr,
+    sync::OnceLock,
+    time::Duration,
+};
+
+use super::{super::PackageSource, PkgCapabilityCache};
+use anyhow::Context;
+use colored::Colorize;
+use dialoguer::theme::ColorfulTheme;
+use virtual_net::{
+    DynVirtualNetworking, IpCidr, IpRoute, NetworkError, Result, StreamSecurity,
+    UnsupportedVirtualNetworking, VirtualIcmpSocket, VirtualNetworking, VirtualRawSocket,
+    VirtualTcpListener, VirtualTcpSocket, VirtualUdpSocket,
+};
+
+/// A custom implementation of the [`virtual_net::VirtualNetwork`] that asks users if they want to
+/// use networking features at runtime.
+#[derive(Debug, Clone)]
+pub(crate) struct AskingNetworking {
+    pkg_cache_path: PathBuf,
+    enable: OnceLock<Result<bool>>,
+    capable: DynVirtualNetworking,
+    unsupported: DynVirtualNetworking,
+}
+
+macro_rules! call {
+    ($self: expr, $fn_name: ident, $( $arg: expr ),* ) => {
+
+        let enable_networking = $self.enable.get_or_init(|| $self.ask_user(stringify!($fn_name)))
+            .map_err(|e| {
+                tracing::error!("{e}");
+                NetworkError::UnknownError
+            })?;
+
+        if enable_networking {
+            return $self.capable.$fn_name( $( $arg ),* ).await;
+        } else {
+            return $self.unsupported.$fn_name( $( $arg ),* ).await;
+        }
+    };
+
+    ($self: expr, $fn_name: ident) => {
+
+        let enable_networking = $self.enable.get_or_init(|| $self.ask_user(stringify!($fn_name))).map_err(|e| {
+                tracing::error!("{e}");
+                NetworkError::UnknownError
+            })?;
+
+        if enable_networking {
+            return $self.capable.$fn_name().await;
+        } else {
+            return $self.unsupported.$fn_name().await;
+        }
+    };
+}
+
+#[derive(Debug, Clone)]
+enum UserSelection {
+    Yes,
+    No,
+    Always,
+}
+
+impl Display for UserSelection {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            UserSelection::Yes => write!(f, "yes"),
+            UserSelection::No => write!(f, "no"),
+            UserSelection::Always => write!(f, "always"),
+        }
+    }
+}
+
+impl FromStr for UserSelection {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let s = s.trim();
+        if s.is_empty() {
+            anyhow::bail!("No input!")
+        }
+
+        let c: char = s.chars().next().unwrap().to_ascii_lowercase();
```

Comment:
Super tiny nit: this stuff is nicer to write without unwraps:

```
let Some(c) = s..trim().chars().next() else {
  bail!(...)
}
```

Although in this case you could just match on the full string as well?

Reply:
> Although in this case you could just match on the full string as well?

I think we want to consider correct many combinations (`y`/`Y`/`yes`/`YES`...). Of course, an input such as `yna` resolves to `yes`